INC-4: add Input

VeraTsitavets opened the pull request:
Добавлен инпут компонент

<img width="878" height="224" alt="Снимок экрана 2025-09-09 153719" src="https://github.com/user-attachments/assets/18a87195-162e-42d6-b66b-ce847a9609d2" />

VeraTsitavets committed:
INC-4 feat: add Input

VeraTsitavets committed:
INC-4 fix: Input isn't a Radix Form component anymore

VeraTsitavets committed:
INC-4 fix: Input's props are ComponentProps<'input'> now, search icon is not a button anymore (it's actually an icon now) & fixed styles

Motya22 commented on src/shared/ui/ControlledInput/ControlledInput.module.scss:
```diff
@@ -0,0 +1,169 @@
+.root {
+  width: 260px;
```
Лучше жёстко не фиксировать ширину, чтобы потом можно было управлять её извне, лучше просто сделать 100%)

Motya22 commented on src/shared/ui/ControlledInput/ControlledInput.module.scss:
```diff
@@ -0,0 +1,169 @@
+.root {
+  width: 260px;
+  margin: 50px 120px;
```
Тоже самое и с внешними отступами, они должны задаваться извне, мы же создаём переиспользуемый компонент и должны быть стили связанные только с самим компонентом непосредственно)
Вот если верстала по BEM, то, например, при создании карточки(блока) мы внешние отступы не задаём, т.к. в разных местах они могут быть разные)

VeraTsitavets replied:
О, ВНЕШНИЕ ОТСТУПЫ ПОПАЛИ ТУДА СЛУЧАЙНО, спасибо что заметил, все учту

Motya22 commented on src/shared/ui/ControlledInput/ControlledInput.module.scss:
```diff
@@ -0,0 +1,169 @@
+
+.label {
+  font-size: 14px;
+  font-weight: 400;
```
400 идёт по-умолчанию, нет смысла задавать)

Motya22 commented on src/shared/ui/ControlledInput/ControlledInput.module.scss:
```diff
@@ -0,0 +1,169 @@
+  align-items: center;
+  justify-content: center;
+
+  font-size: 16px;
```
16px тоже по-умолчанию в браузере)

Motya22 commented on src/shared/assets/svg/outline/eye-off-outline.svg:
```diff
@@ -0,0 +1,12 @@
+<svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
+<g clip-path="url(#clip0_301_3928)">
+<path d="M4.71 3.29C4.61676 3.19676 4.50607 3.1228 4.38425 3.07234C4.26243 3.02188 4.13186 2.99591 4 2.99591C3.86814 2.99591 3.73758 3.02188 3.61575 3.07234C3.49393 3.1228 3.38324 3.19676 3.29 3.29C3.1017 3.47831 2.99591 3.7337 2.99591 4C2.99591 4.2663 3.1017 4.5217 3.29 4.71L8.92 10.34C8.56381 11.0026 8.43059 11.7624 8.54009 12.5067C8.64959 13.2509 8.99599 13.9401 9.52794 14.4721C10.0599 15.004  ...
```
Везде в "<path d=" нужно заменить fill="white" на fill="currentColor", чтобы можно было менять цвет извне)

Motya22 commented on src/shared/ui/ControlledInput/ControlledInput.module.scss:
```diff
@@ -0,0 +1,169 @@
+
+  width: 34px;
+  height: 34px;
+  margin-right: 12px;
```
Зачем margin-right? ты же абсолютно позиционируешь и можешь с помощью right этот отступ задать)

Motya22 commented on src/shared/ui/ControlledInput/ControlledInput.module.scss:
```diff
@@ -0,0 +1,169 @@
+  position: absolute;
+  transform: translateY(-50%);
+
+  width: 34px;
```
Не совсем понимаю, почему у кнопки размеры больше, чем у иконки, по макету 24px)

Motya22 commented on src/shared/ui/ControlledInput/ControlledInput.module.scss:
```diff
@@ -0,0 +1,169 @@
+  &.eyeButton{
+    top: 21px;
+    right: 0;
+    bottom: 50%;
```
Нужно top 50%, а не bottom и right 12px)

Motya22 commented on src/shared/ui/ControlledInput/ControlledInput.module.scss:
```diff
@@ -0,0 +1,169 @@
+  }
+
+  &.search {
+    padding-left: 40px;
```
Ещё нужно padding-right задать, когда инпут-пароль, чтобы текст под иконкой не был)

Motya22 commented on src/shared/ui/ControlledInput/ControlledInput.module.scss:
```diff
@@ -0,0 +1,169 @@
+  }
+
+  &.search{
+    top: 18px;
```
Здесь тоже top: 50%, а left 12px, bottom не нужен)

Motya22 commented on src/shared/ui/ControlledInput/ControlledInput.module.scss:
```diff
@@ -0,0 +1,169 @@
+
+    height: 20px;
+    width: 20px;
+    margin: 0 9px 0 12px;
```
matgin не нужен, если сделаешь как я написал выше)

Motya22 commented on src/shared/ui/ControlledInput/ControlledInput.tsx:
```diff
@@ -0,0 +1,96 @@
+type Props = ComponentProps<'input'> & {
+  errorMessage?: string
+  label?: string
+  disabled?: boolean
```
Зачем переопределять стандартные пропсы инпута(disabled, placeholder, type)?)

Motya22 commented on src/shared/ui/ControlledInput/ControlledInput.tsx:
```diff
@@ -0,0 +1,96 @@
+  errorMessage?: string
+  label?: string
+  disabled?: boolean
+  onValueChange?: (value: string) => void
```
Лучше оставь стандартный onChange, меньше проблем будет)

Motya22 commented on src/shared/ui/ControlledInput/ControlledInput.tsx:
```diff
@@ -0,0 +1,96 @@
+  onChange,
+  type,
+  ref,
+  ...props
```
Достань стандартный className и передай в конец clsx для инпута, чтобы можно было стилизовать извне)

Motya22 commented on src/shared/ui/ControlledInput/ControlledInput.tsx:
```diff
@@ -0,0 +1,96 @@
+}: Props) {
+  const [showPassword, setShowPassword] = useState(false)
+
+  const endTypeCalc = (elementType: string, passwordVisible: boolean) => {
```
Вынеси эту утилитную функцию за компонент, чтобы она не создавалась каждый раз при ререндере или вообще прям инлайном в инпуте вычисляй: "type={isPasswordType && isPasswordVisible ? 'text' : type}"

Motya22 commented on src/shared/ui/ControlledInput/ControlledInput.tsx:
```diff
@@ -0,0 +1,96 @@
+    return elementType
+  }
+
+  const showPasswordToggle = type === 'password'
```
я бы переименовал в флаги isPasswordType и isSearchType)

Motya22 commented on src/shared/ui/ControlledInput/ControlledInput.tsx:
```diff
@@ -0,0 +1,96 @@
+
+  const endType = endTypeCalc(type, showPassword)
+
+  function handleChange(e: ChangeEvent<HTMLInputElement>) {
```
Как я уже писал выше, лучше не переопределяй стандартный onChange)

Motya22 commented on src/shared/ui/ControlledInput/ControlledInput.tsx:
```diff
@@ -0,0 +1,96 @@
+    onChange?.(e)
+  }
+
+  const classNames = {
```
По поводу объекта со стилями вкусовщина, но я бы прям инлайн вычислял)

Motya22 commented on src/shared/ui/ControlledInput/ControlledInput.tsx:
```diff
@@ -0,0 +1,96 @@
+      type === 'search' && styles.search
+    ),
+    label: clsx(styles.label, !!errorMessage && styles.error, disabled && styles.disabled),
+    icon: clsx(styles.icon, !!errorMessage && styles.error, disabled && styles.disabled),
```
я не понимаю, почему везде дублируется "!!errorMessage && styles.error, disabled && styles.disabled". Ты же можешь отталкиваться от класса/состояния, который висит на родителе в стилях)

Motya22 commented on src/shared/ui/ControlledInput/ControlledInput.tsx:
```diff
@@ -0,0 +1,96 @@
+  return (
+    <div className={classNames.root}>
+      {label && <label className={classNames.label}>{label}</label>}
+      <div className={clsx(styles.divContainer)}>
```
Здесь clsx не нужен и определись, использовать объект classNames или инлайново, старайся писать с одном стиле)

Motya22 commented on src/shared/ui/ControlledInput/ControlledInput.tsx:
```diff
@@ -0,0 +1,96 @@
+      {label && <label className={classNames.label}>{label}</label>}
+      <div className={clsx(styles.divContainer)}>
+        <input
+          onChange={() => handleChange}
```
Это не будет работать, нужно было либо просто передать onChange={handleChange}, либо "onChange={(e) => handleChange(e)}".
Но как я уже говорил, лучше не переопределяй стандартный onChange и он пойдёт в инпут из "{...props}")

Motya22 commented on src/shared/ui/ControlledInput/ControlledInput.tsx:
```diff
@@ -0,0 +1,96 @@
+      styles.input,
+      !!errorMessage && styles.error,
+      disabled && styles.disabled,
+      type === 'search' && styles.search
```
Вот здесь в конец нужно добавить стандартный className)

Motya22 commented on src/shared/ui/ControlledInput/ControlledInput.stories.tsx:
```diff
@@ -0,0 +1,57 @@
+  parameters: {
+    docs: {
+      description: {
+        component: 'Управляемый инпут на базе Radix UI.',
```
Он не на базе Radix UI, ты же его не используешь, а сама сделала на базе стандартного инпута)

Motya22 commented on src/shared/ui/ControlledInput/ControlledInput.stories.tsx:
```diff
@@ -0,0 +1,57 @@
+    errorMessage: 'wrong password',
+    label: 'Password',
+    placeholder: 'password1234',
+    type: "'password'",
```
Из-за того, что здесь и дальше по коду двое разных кавычек, не работает, оставь одни)

Motya22 commented on src/shared/ui/ControlledInput/ControlledInput.stories.tsx:
```diff
@@ -0,0 +1,57 @@
+  },
+}
+
+export const DefaultEmail: Story = {
```
У меня в последней истории почему-то такая ошибка в WebStorm(:
<img width="1920" height="1200" alt="image" src="https://github.com/user-attachments/assets/1d7de5ea-d206-4207-8dac-bb5233e6f5da" />

Motya22 reviewed (changes requested):
В целом, молодец, я по-максимуму сказал, что можно улучшить)
Можешь на поддержку сгонять, ИИ попросить провести ревью)

VeraTsitavets committed:
INC-4 fix: fix minor mistakes in ControledInput scss styles

&& delete duplicating props,
minor fix of names and functions,
rewrite type changing function

&& fix type mistakes in storybook
fix error in the Default story

VeraTsitavets committed:
INC-4 refactor: refactor of the classNames

Motya22 commented on src/shared/ui/ControlledInput/ControlledInput.tsx:
```diff
@@ -10,85 +10,80 @@ import styles from './ControlledInput.module.scss'
-  disabled?: boolean
-  onValueChange?: (value: string) => void
-  placeholder?: string
+  onChange?: (value: string) => void
```
onChange это тоже стандартный проп, его не нужно переопределять)

Motya22 commented on src/shared/ui/ControlledInput/ControlledInput.tsx:
```diff
@@ -10,85 +10,80 @@ import styles from './ControlledInput.module.scss'
+  const isPasswordType = type === 'password'
+  const isSearchType = type === 'search'
 
   function handleChange(e: ChangeEvent<HTMLInputElement>) {
```
Эту функцию удаляй)

Motya22 commented on src/shared/ui/ControlledInput/ControlledInput.tsx:
```diff
@@ -10,85 +10,80 @@ import styles from './ControlledInput.module.scss'
-          onChange={() => handleChange}
-          className={classNames.input}
-          type={endType}
+          onChange={handleChange}
```
Здесь onChange тоже удаляй и стандартный попадёт в "...props" и передастся через "{...props}" инпуту)

Motya22 commented on src/shared/ui/ControlledInput/ControlledInput.tsx:
```diff
@@ -10,85 +10,80 @@ import styles from './ControlledInput.module.scss'
   label,
   disabled,
-  onValueChange,
   placeholder,
```
Не доставай отсюда placeholder)

Motya22 commented on src/shared/ui/ControlledInput/ControlledInput.tsx:
```diff
@@ -10,85 +10,80 @@ import styles from './ControlledInput.module.scss'
+            className
+          )}
+          type={isPasswordType && showPassword ? 'text' : type}
           placeholder={placeholder}
```
Эту строчку удаляй и тогда placeholder тоже передастся через "{...props}", смысл этот стандартный проп доставать и вручную передавать, если ты его нигде больше в компоненте не используешь, например, как проп disabled для стилизации)

Motya22 commented on src/shared/ui/ControlledInput/ControlledInput.tsx:
```diff
@@ -10,85 +10,80 @@ import styles from './ControlledInput.module.scss'
           >
             {showPassword ? (
-              <EyeOpenIcon className={classNames.icon} />
+              <EyeOpenIcon className={clsx(styles.icon)} />
```
Здесь не нужен clsx, он нужен там, где мы хотим несколько классов присвоить)

Motya22 commented on src/shared/ui/ControlledInput/ControlledInput.tsx:
```diff
@@ -10,85 +10,80 @@ import styles from './ControlledInput.module.scss'
+              <EyeOpenIcon className={clsx(styles.icon)} />
             ) : (
-              <EyeClosedIcon className={classNames.icon} />
+              <EyeClosedIcon className={clsx(styles.icon)} />
```
Здесь тоже не нужен clsx)

Motya22 commented on src/shared/ui/ControlledInput/ControlledInput.tsx:
```diff
@@ -10,85 +10,80 @@ import styles from './ControlledInput.module.scss'
+        )}
       </div>
-      {!!errorMessage && <span className={classNames.message}>{errorMessage}</span>}
+      {!!errorMessage && <span className={clsx(styles.errorMessage)}>{errorMessage}</span>}
```
И здесь не нужен clsx)

Motya22 reviewed (commented):
Исправляй и мержи ПР тогда)

VeraTsitavets committed:
INC-4 refactor: refactor of ControlledInput
